### PR TITLE
Gadams3/fix material builder dependencies

### PR DESCRIPTION
**Pull request by gadams3**
## What does this PR do?

- Changed the material and material type builders to better handle intermediate asset dependencies.
- Changed the material builder to directly assign absolute and wild card job dependencies for generated material types and shaders. This will prevent the material builder from processing until all of the shaders generated material types and shaders have completed their own processing. This helps enforce that the material updates and reloads after material type and shader changes. This also eliminates numerous warnings about unresolved or missing dependencies.
- Updated the abstract material type dependency to use the common platform id correctly.
- Updated material and material type source data to use the complete resolve file name in log messages.

## How was this PR tested?

- Verified that all assets build in the automated testing project using server and pc configurations on the asset profile build.
- Verified that all assets build in the multiplayer sample project on Windows PC build.
- Verified that materials and material types no longer report several warnings or errors on the above projects.
- Verified that material assets process after generated shaders and material types and the hot reload in the main editor, at least as much as extensive  manual testing confirmed.

### CODE REVIEW

File: Gems/Atom/RPI/Code/Include/Atom/RPI.Edit/Material/MaterialTypeSourceData.h [hunk 1]
```diff
@@ -299,6 +299,9 @@ namespace AZ
             //! See enum Format
             Format GetFormat() const;
 
+            //! Return a concatenated list of shader references from mall collections
```
Comment (AMZN-alexpete) [inline Gems/Atom/RPI/Code/Include/Atom/RPI.Edit/Material/MaterialTypeSourceData.h:302]:
```suggestion
            //! Return a concatenated list of shader references from all collections
```

**Review by lsemp3d — APPROVED** [review]

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp [hunk 2]
```diff
@@ -136,51 +137,76 @@ namespace AZ
...
-                    const AZStd::string intermediateMaterialTypePath =
-                        MaterialUtils::PredictIntermediateMaterialTypeSourcePath(materialSourcePath, resolvedMaterialTypePath);
+                    // Create a dependency on the abstract, pipeline, version of the material type and its products. The pipeline based
+                    // material type builder uses the 'common' asset platform ID because it produces immediate assets. The sub ID filter
+                    // should remain empty to observe all produced intermediate assets.
+                    auto& materialTypeJobDependency = outputJobDescriptor.m_jobDependencyList.emplace_back();
+                    materialTypeJobDependency.m_type = AssetBuilderSDK::JobDependencyType::Order;
+                    materialTypeJobDependency.m_sourceFile.m_sourceFileDependencyPath = resolvedMaterialTypePath;
+                    materialTypeJobDependency.m_jobKey = MaterialTypeBuilder::PipelineStageJobKey;
+                    materialTypeJobDependency.m_platformIdentifier = AssetBuilderSDK::CommonPlatformName;
```
Comment (AMZN-alexpete) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:168]: my eyes gloss over when looking at all these blocks of code.  I wish we had a simpler way to express a dependency e.g.
````
auto materialTypeJobDependency = OrderDependency( resolvedMaterialTypePath, PipelinStageJobKey );
```

or
```
auto materialTypeJobDependency {
    resolvedMaterialTypePath,
    PipelinStageJobKey,
    CommonPlatformName,
    JobDependencyType::Order
};
```

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp [hunk 3]
```diff
@@ -234,15 +260,15 @@ namespace AZ
             const auto& materialSourceData = materialSourceDataOutcome.GetValue();
 
             // Load the material file and create the MaterialAsset object
-            auto materialAssetOutcome = materialSourceData.CreateMaterialAsset(
+            const auto& materialAssetOutcome = materialSourceData.CreateMaterialAsset(
```
Comment (AMZN-alexpete) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:263]: `CreateMaterialAsset` does not return a const ref

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp (hunk 2 above)
Comment (gadams3) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:168]: That is part of what the previous function utility function was supposed to do but it did a lot of extra guesswork trying to compensate for missing or yet to be created intermediate assets. I could make a function that does just this for all the cases that follow the exact pattern.

File: Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialSourceData.cpp [hunk 4]
```diff
@@ -141,69 +140,66 @@ namespace AZ
...
-
             materialAssetCreator.SetMaterialTypeVersion(m_materialTypeVersion);
 
             if (!m_parentMaterial.empty())
             {
-                constexpr uint32_t subId = 0;
-                auto parentMaterialAsset = AssetUtils::LoadAsset<MaterialAsset>(
-                    materialSourceFilePath, m_parentMaterial, subId, AssetUtils::TraceLevel::Error, dontLoadImageAssets);
-                if (!parentMaterialAsset.IsSuccess())
+                const auto& parentMaterialAssetOutcome = AssetUtils::LoadAsset<MaterialAsset>(
```
Comment (AMZN-alexpete) [inline Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialSourceData.cpp:167]: ditto here regarding const ref

File: Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialSourceData.cpp [hunk 5]
```diff
@@ -232,25 +226,26 @@ namespace AZ
                 return Failure();
             }
 
-            const AZStd::string materialTypeSourcePath = MaterialUtils::GetFinalMaterialTypeSourcePath(materialSourceFilePath, m_materialType);
-            const auto materialTypeAssetId = MaterialUtils::GetFinalMaterialTypeAssetId(materialSourceFilePath, m_materialType);
+            const auto& materialTypeSourcePath = MaterialUtils::GetFinalMaterialTypeSourcePath(materialSourceFilePath, m_materialType);
```
Comment (AMZN-alexpete) [inline Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialSourceData.cpp:229]: `GetFinalMaterialTypeSourcePath` does not return a const ref

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp (hunk 2 above)
Comment (AMZN-alexpete) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:168]: I'm fine with that change being in another PR, I'd rather get the fix in, then delay it

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp (hunk 3 above)
Comment (gadams3) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Material/MaterialBuilder.cpp:263]: Old habit. It does not return one but avoids copying, which probably is not necessary anymore, by extending the lifetime of a referenced temporary.

**Review by AMZN-alexpete — APPROVED** [review]
I'd prefer not to use const ref variables for functions that don't return a const ref to avoid any assumptions future devs who work in this code may make about usage, ownership and return signature of the functions involved.

**Review by moudgils — APPROVED** [review]

**Pull request merged**